source/_posts/2016-10-24-gsoc-2016-port-nmatrix-to-jruby.markdown

```diff
@@ -0,0 +1,536 @@
+      nmatrix.s = ArrayRealVector.new result
+
+      return nmatrix
+    else
+      return self.solve rhs
+    end
+  end
+```
+
+Currently, Hessenberg transformation for NMatrix-JRuby has not been implemented.
+
+## **Other dtypes**
+
+We have tried implementing float dtypes using jblas FloatMatrix.  jblas was used instead of Commons Math as the latter uses Field Elements for Floats and it had some issues with Reflection and TypeErasure. However, it resulted in errors due to precision.
+
+
+## **Code Organisation and Deployment**
+
+To minimise conflict with the MRI codebase all the jruby front end code has been placed in `/lib/nmatrix/jruby` directory. `lib/nmatrix/nmatrix.rb` decides whether to load `nmatrix.so` or `nmatrix_jruby.rb` after detecting the Ruby Platform.
+
+The added advantage is that the ruby interpreter must not decide which function to call at run-time. The impact on performance can be seen when programs which intensively use NMatrix for linear algebraic computations(e.g. mixed-models) are run.
+
+## **Test Report**
+
+After the port; this is the final report that summarizes the number of tests that successfully pass:
+
+**NMatrix**
+
+|Spec file|Total Tests|Success|Failure|Pending|
+|------------|:------------:|:-----------:|:-------------:|:-------------:|
+|00_nmatrix_spec|188|139|43|6|
+|01_enum_spec|17|8|09|0|
+|02_slice_spec|144|116|28|0|
+|03_nmatrix_monkeys_spec|12|11|01|0|
+|elementwise_spec|38|21|17|0|
+|homogeneous_spec.rb|07|06|01|0|
+|math_spec|737|541|196|0|
+|shortcuts_spec|81|57|24|0|
+|stat_spec|72|40|32|0|
+|slice_set_spec|6|2|04|0|
+
+<br>
+Why some tests fail?
+
+1.  Complex dtype has not been implemented.
+2.  Sparse matrices (list and yale) have not been implemented.
+3.  Decomposition methods that are specific to LAPACK and ATLAS have not been implemented.
+4.  Integer dtype is not properly assigned to Floor, Ceil and Round methods.
+
+**Mixed-Models**
+
+|Spec file|Total Test|Success|Failure|Pending|
+|------------|:------------:|:-----------:|:-------------:|:-------------:|
+|Deviance_spec|04|04|0|0|
+|LMM_spec|195|195|0|0|
+|LMM_categorical_data_spec.rb|48|45|3|0|
+|LMMFormula_spec.rb|05|05|0|0|
+|LMM_interaction_effects_spec.rb|82|82|0|0|
+|LMM_nested_effects_spec.rb|40|40|0|0|
+|matrix_methods_spec.rb|52|52|0|0|
+|ModelSpecification_spec.rb|07|07|0|0|
+|NelderMeadWithConstraints_spec.rb|08|08|0|0|
+
+<br>
+
+
+## **Future work**
+
+NMatrix on JRuby offers comparable speeds to MRI. For specific computations it will be possible to leverage the threading support of JRuby and speed up things using multiple cores.
+
+Adding new functionality to NMatrix-JRuby will be easy from here. Personally, I am interested to add OpenCL support to leverage the GPU computational capacity available on most machines today.
+
+## **Conclusion**
+The main goal of this project was to bring  **Scientific Computation to JRuby**, to gain from the performance JRuby offers.
```

Comment:
I'm going to pick on this a bit. Doesn't JRuby already have scientific computation? I'd think the project was more about providing a unified interface for linear algebra between MRI and JRuby. Are you saying use of JRuby and Apache Commons Math can speed up NMatrix over MRI?